README.md
[![License](https://img.shields.io/badge/Code_License-Modified_MIT-blue.svg)](LICENSE)
[![License](https://img.shields.io/badge/Weights_License-DeepFloyd_IF-orange.svg)](LICENSE-MODEL)
[![Downloads](https://pepy.tech/badge/deepfloyd_if)](https://pepy.tech/project/deepfloyd_if)
[![Discord](https://img.shields.io/badge/Discord-%237289DA.svg?logo=discord&logoColor=white)](https://discord.gg/umz62Mgr)
[![Twitter](https://img.shields.io/badge/Twitter-%231DA1F2.svg?logo=twitter&logoColor=white)](https://twitter.com/deepfloydai)
[![Linktree](https://img.shields.io/badge/Linktree-%2339E09B.svg?logo=linktree&logoColor=white)](http://linktr.ee/deepfloyd)

# IF by [DeepFloyd Lab](https://deepfloyd.ai) at [StabilityAI](https://stability.ai/)

<p align="center">
  <img src="./pics/nabla.jpg" width="100%">
</p>

We introduce DeepFloyd IF, a novel state-of-the-art open-source text-to-image model with a high degree of photorealism and language understanding. DeepFloyd IF is a modular composed of a frozen text encoder and three cascaded pixel diffusion modules: a base model that generates 64x64 px image based on text prompt and two super-resolution models, each designed to generate images of increasing resolution: 256x256 px and 1024x1024 px. All stages of the model utilize a frozen text encoder based on the T5 transformer to extract text embeddings, which are then fed into a UNet architecture enhanced with cross-attention and attention pooling. The result is a highly efficient model that outperforms current state-of-the-art models, achieving a zero-shot FID score of 6.66 on the COCO dataset. Our work underscores the potential of larger UNet architectures in the first stage of cascaded diffusion models and depicts a promising future for text-to-image synthesis.

<p align="center">
  <img src="./pics/deepfloyd_if_scheme.jpg" width="100%">
</p>

*Inspired by* [*Photorealistic Text-to-Image Diffusion Models with Deep Language Understanding*](https://arxiv.org/pdf/2205.11487.pdf)

## Minimum requirements to use all IF models:
- 16GB vRAM for IF-I-XL (4.3B text to 64x64 base module) & IF-II-L (1.2B to 256x256 upscaler module)
- 24GB vRAM for IF-I-XL (4.3B text to 64x64 base module) & IF-II-L (1.2B to 256x256 upscaler module) & Stable x4 (to 1024x1024 upscaler)
- `xformers` and set env variable `FORCE_MEM_EFFICIENT_ATTN=1`


## Quick Start
[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/huggingface/notebooks/blob/main/diffusers/deepfloyd_if_free_tier_google_colab.ipynb)
[![Hugging Face Spaces](https://img.shields.io/badge/%F0%9F%A4%97%20Hugging%20Face-Spaces-blue)](https://huggingface.co/spaces/DeepFloyd/IF)

```shell
pip install deepfloyd_if==1.0.2rc1
pip install xformers==0.0.16
pip install git+https://github.com/openai/CLIP.git --no-deps
```

## Local notebooks
[![Jupyter Notebook](https://img.shields.io/badge/jupyter_notebook-%23FF7A01.svg?logo=jupyter&logoColor=white)](https://huggingface.co/DeepFloyd/IF-notebooks/blob/main/pipes-DeepFloyd-IF-v1.0.ipynb)
[![Kaggle](https://kaggle.com/static/images/open-in-kaggle.svg)](https://www.kaggle.com/code/shonenkov/deepfloyd-if-4-3b-generator-of-pictures)

The Dream, Style Transfer, Super Resolution or Inpainting modes are avaliable in a Jupyter Notebook [here](https://huggingface.co/DeepFloyd/IF-notebooks/blob/main/pipes-DeepFloyd-IF-v1.0.ipynb).



## Integration with 🤗 Diffusers

IF is also integrated with the 🤗 Hugging Face [Diffusers library](https://github.com/huggingface/diffusers/).

Diffusers runs each stage individually allowing the user to customize the image generation process as well as allowing to inspect intermediate results easily.

### Example

Before you can use IF, you need to accept its usage conditions. To do so:
1. Make sure to have a [Hugging Face account](https://huggingface.co/join) and be loggin in
2. Accept the license on the model card of [DeepFloyd/IF-I-XL-v1.0](https://huggingface.co/DeepFloyd/IF-I-XL-v1.0)
3. Make sure to login locally. Install `huggingface_hub`
```sh
pip install huggingface_hub --upgrade
```

run the login function in a Python shell

```py
from huggingface_hub import login

login()
```

and enter your [Hugging Face Hub access token](https://huggingface.co/docs/hub/security-tokens#what-are-user-access-tokens).

Next we install `diffusers` and dependencies:

```sh
pip install diffusers accelerate transformers safetensors
```

And we can now run the model locally.

By default `diffusers` makes use of [model cpu offloading](https://huggingface.co/docs/diffusers/optimization/fp16#model-offloading-for-fast-inference-and-memory-savings) to run the whole IF pipeline with as little as 14 GB of VRAM.

If you are using `torch>=2.0.0`, make sure to **delete all** `enable_xformers_memory_efficient_attention()`
functions.

```py
from diffusers import DiffusionPipeline
from diffusers.utils import pt_to_pil
import torch

# stage 1
stage_1 = DiffusionPipeline.from_pretrained("DeepFloyd/IF-I-XL-v1.0", variant="fp16", torch_dtype=torch.float16)
stage_1.enable_xformers_memory_efficient_attention()  # remove line if torch.__version__ >= 2.0.0
stage_1.enable_model_cpu_offload()

# stage 2
stage_2 = DiffusionPipeline.from_pretrained(
    "DeepFloyd/IF-II-L-v1.0", text_encoder=None, variant="fp16", torch_dtype=torch.float16
)
stage_2.enable_xformers_memory_efficient_attention()  # remove line if torch.__version__ >= 2.0.0
stage_2.enable_model_cpu_offload()

# stage 3
safety_modules = {"feature_extractor": stage_1.feature_extractor, "safety_checker": stage_1.safety_checker, "watermarker": stage_1.watermarker}
stage_3 = DiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-x4-upscaler", **safety_modules, torch_dtype=torch.float16)
stage_3.enable_xformers_memory_efficient_attention()  # remove line if torch.__version__ >= 2.0.0
stage_3.enable_model_cpu_offload()

prompt = 'a photo of a kangaroo wearing an orange hoodie and blue sunglasses standing in front of the eiffel tower holding a sign that says "very deep learning"'

# text embeds
prompt_embeds, negative_embeds = stage_1.encode_prompt(prompt)

generator = torch.manual_seed(0)

# stage 1
image = stage_1(prompt_embeds=prompt_embeds, negative_prompt_embeds=negative_embeds, generator=generator, output_type="pt").images
pt_to_pil(image)[0].save("./if_stage_I.png")

# stage 2
image = stage_2(
    image=image, prompt_embeds=prompt_embeds, negative_prompt_embeds=negative_embeds, generator=generator, output_type="pt"
).images
pt_to_pil(image)[0].save("./if_stage_II.png")

# stage 3
image = stage_3(prompt=prompt, image=image, generator=generator, noise_level=100).images
image[0].save("./if_stage_III.png")
```

 There are multiple ways to speed up the inference time and lower the memory consumption even more with `diffusers`. To do so, please have a look at the Diffusers docs:

- 🚀 [Optimizing for inference time](https://huggingface.co/docs/diffusers/api/pipelines/if#optimizing-for-speed)
- ⚙️ [Optimizing for low memory during inference](https://huggingface.co/docs/diffusers/api/pipelines/if#optimizing-for-memory)

For more in-detail information about how to use IF, please have a look at [the IF blog post](https://huggingface.co/blog/if) and [the documentation](https://huggingface.co/docs/diffusers/main/en/api/pipelines/if) 📖.

Diffusers dreambooth scripts also supports fine-tuning 🎨 [IF](https://huggingface.co/docs/diffusers/main/en/training/dreambooth#if).
With parameter efficient finetuning, you can add new concepts to IF with a single GPU and ~28 GB VRAM.

## Run the code locally

### Loading the models into VRAM

```python
from deepfloyd_if.modules import IFStageI, IFStageII, StableStageIII
from deepfloyd_if.modules.t5 import T5Embedder

device = 'cuda:0'
if_I = IFStageI('IF-I-XL-v1.0', device=device)
if_II = IFStageII('IF-II-L-v1.0', device=device)
if_III = StableStageIII('stable-diffusion-x4-upscaler', device=device)
t5 = T5Embedder(device="cpu")
```

### I. Dream
Dream is the text-to-image mode of the IF model

```python
from deepfloyd_if.pipelines import dream

prompt = 'ultra close-up color photo portrait of rainbow owl with deer horns in the woods'
count = 4

result = dream(
    t5=t5, if_I=if_I, if_II=if_II, if_III=if_III,
    prompt=[prompt]*count,
    seed=42,
    if_I_kwargs={
        "guidance_scale": 7.0,
        "sample_timestep_respacing": "smart100",
    },
    if_II_kwargs={
        "guidance_scale": 4.0,
        "sample_timestep_respacing": "smart50",
    },
    if_III_kwargs={
        "guidance_scale": 9.0,
        "noise_level": 20,
        "sample_timestep_respacing": "75",
    },
)

if_III.show(result['III'], size=14)
```
![](./pics/dream-III.jpg)

## II. Zero-shot Image-to-Image Translation

![](./pics/img_to_img_scheme.jpeg)

In Style Transfer mode, the output of your prompt comes out at the style of the `support_pil_img`
```python
from deepfloyd_if.pipelines import style_transfer

result = style_transfer(
    t5=t5, if_I=if_I, if_II=if_II,
    support_pil_img=raw_pil_image,
    style_prompt=[
        'in style of professional origami',
        'in style of oil art, Tate modern',
        'in style of plastic building bricks',
        'in style of classic anime from 1990',
    ],
    seed=42,
    if_I_kwargs={
        "guidance_scale": 10.0,
        "sample_timestep_respacing": "10,10,10,10,10,10,10,10,0,0",
        'support_noise_less_qsample_steps': 5,
    },
    if_II_kwargs={
        "guidance_scale": 4.0,
        "sample_timestep_respacing": 'smart50',
        "support_noise_less_qsample_steps": 5,
    },
)
if_I.show(result['II'], 1, 20)
```

![Alternative Text](./pics/deep_floyd_if_image_2_image.gif)


## III. Super Resolution
For super-resolution, users can run `IF-II` and `IF-III` or 'Stable x4' on an image that was not necessarely generated by IF (two cascades):

```python
from deepfloyd_if.pipelines import super_resolution

middle_res = super_resolution(
    t5,
    if_III=if_II,
    prompt=['woman with a blue headscarf and a blue sweaterp, detailed picture, 4k dslr, best quality'],
    support_pil_img=raw_pil_image,
    img_scale=4.,
    img_size=64,
    if_III_kwargs={
        'sample_timestep_respacing': 'smart100',
        'aug_level': 0.5,
        'guidance_scale': 6.0,
    },
)
high_res = super_resolution(
    t5,
    if_III=if_III,
    prompt=[''],
    support_pil_img=middle_res['III'][0],
    img_scale=4.,
    img_size=256,
    if_III_kwargs={
        "guidance_scale": 9.0,
        "noise_level": 20,
        "sample_timestep_respacing": "75",
    },
)
show_superres(raw_pil_image, high_res['III'][0])
```

![](./pics/if_as_upscaler.jpg)


### IV. Zero-shot Inpainting

```python
from deepfloyd_if.pipelines import inpainting

result = inpainting(
    t5=t5, if_I=if_I,
    if_II=if_II,
    if_III=if_III,
    support_pil_img=raw_pil_image,
    inpainting_mask=inpainting_mask,
    prompt=[
        'oil art, a man in a hat',
    ],
    seed=42,
    if_I_kwargs={
        "guidance_scale": 7.0,
        "sample_timestep_respacing": "10,10,10,10,10,0,0,0,0,0",
        'support_noise_less_qsample_steps': 0,
    },
    if_II_kwargs={
        "guidance_scale": 4.0,
        'aug_level': 0.0,
        "sample_timestep_respacing": '100',
    },
    if_III_kwargs={
        "guidance_scale": 9.0,
        "noise_level": 20,
        "sample_timestep_respacing": "75",
    },
)
if_I.show(result['I'], 2, 3)
if_I.show(result['II'], 2, 6)
if_I.show(result['III'], 2, 14)
```
![](./pics/deep_floyd_if_inpainting.gif)

### 🤗 Model Zoo 🤗
The link to download the weights as well as the model cards will be available soon on each model of the model zoo

#### Original

| Name                                                      | Cascade | Params | FID  | Batch size | Steps |
|:----------------------------------------------------------|:-------:|:------:|:----:|:----------:|:-----:|
| [IF-I-M](https://huggingface.co/DeepFloyd/IF-I-M-v1.0)    |    I    |  400M  | 8.86 |    3072    | 2.5M  |
| [IF-I-L](https://huggingface.co/DeepFloyd/IF-I-L-v1.0)    |    I    |  900M  | 8.06 |    3200    | 3.0M  |
| [IF-I-XL](https://huggingface.co/DeepFloyd/IF-I-XL-v1.0)* |    I    |  4.3B  | 6.66 |    3072    | 2.42M |
| [IF-II-M](https://huggingface.co/DeepFloyd/IF-II-M-v1.0)  |   II    |  450M  |  -   |    1536    | 2.5M  |
| [IF-II-L](https://huggingface.co/DeepFloyd/IF-II-L-v1.0)* |   II    |  1.2B  |  -   |    1536    | 2.5M  |
| IF-III-L* _(soon)_                                        |   III   |  700M  |  -   |    3072    | 1.25M |

 *best modules

### Quantitative Evaluation

`FID = 6.66`

![](./pics/fid30k_if.jpg)

## License

The code in this repository is released under the bespoke license (see added [point two](https://github.com/deep-floyd/IF/blob/main/LICENSE#L13)).

The weights will be available soon via [the DeepFloyd organization at Hugging Face](https://huggingface.co/DeepFloyd) and have their own LICENSE.

**Disclaimer:** *The initial release of the IF model is under a restricted research-purposes-only license temporarily to gather feedback, and after that we intend to release a fully open-source model in line with other Stability AI models.*

## Limitations and Biases

The models available in this codebase have known limitations and biases. Please refer to [the model card](https://huggingface.co/DeepFloyd/IF-I-L-v1.0) for more information.


## 🎓 DeepFloyd IF creators:

- Alex Shonenkov [GitHub](https://github.com/shonenkov) | [Linktr](https://linktr.ee/shonenkovAI)
- Misha Konstantinov [GitHub](https://github.com/zeroshot-ai) | [Twitter](https://twitter.com/_bra_ket)
- Daria Bakshandaeva [GitHub](https://github.com/Gugutse) | [Twitter](https://twitter.com/_gugutse_)
- Christoph Schuhmann [GitHub](https://github.com/christophschuhmann) | [Twitter](https://twitter.com/laion_ai)
- Ksenia Ivanova [GitHub](https://github.com/ivksu) | [Twitter](https://twitter.com/susiaiv)
- Nadiia Klokova [GitHub](https://github.com/vauimpuls) | [Twitter](https://twitter.com/vauimpuls)


## 📄 Research Paper (Soon)

## Acknowledgements

Special thanks to [StabilityAI](http://stability.ai) and its CEO [Emad Mostaque](https://twitter.com/emostaque) for invaluable support, providing GPU compute and infrastructure to train the models (our gratitude goes to [Richard Vencu](https://github.com/rvencu)); thanks to [LAION](https://laion.ai) and [Christoph Schuhmann](https://github.com/christophschuhmann) in particular for contribution to the project and well-prepared datasets; thanks to [Huggingface](https://huggingface.co) teams for optimizing models' speed and memory consumption during inference, creating demos and giving cool advice!

## 🚀 External Contributors 🚀
- The Biggest Thanks [@Apolinário](https://github.com/apolinario), for ideas, consultations, help and support on all stages to make IF available in open-source; for writing a lot of documentation and instructions; for creating a friendly atmosphere in difficult moments 🦉;
- Thanks, [@patrickvonplaten](https://github.com/patrickvonplaten), for improving loading time of unet models by 80%;
for integration Stable-Diffusion-x4 as native pipeline 💪;
- Thanks, [@williamberman](https://github.com/williamberman) and [@patrickvonplaten](https://github.com/patrickvonplaten) for diffusers integration 🙌;
- Thanks, [@hysts](https://github.com/hysts) and [@Apolinário](https://github.com/apolinario) for creating [the best gradio demo with IF](https://huggingface.co/spaces/DeepFloyd/IF) 🚀;
- Thanks, [@Dango233](https://github.com/Dango233), for adapting IF with xformers memory efficient attention 💪;


CHANGELOG.md
v1.1.0rc
-------

- uses separated tokenizer_path to init tokenizer in T5Embedder
- pytorch2 support
- more effective attention
- freeU support, adding more details
- SD-upscaler: fix normalization for new version of diffusers

v1.0.1
------

- renamed main model `IF-I-IF` --> `IF-I-XL`
- moved dir `notebooks` to HF storage https://huggingface.co/DeepFloyd/IF-notebooks; lets keep new notebooks there;
- added additional kaggle notebook (more free GPU resources) how to generate pictures 1k: [![Kaggle](https://kaggle.com/static/images/open-in-kaggle.svg)](https://www.kaggle.com/code/shonenkov/deepfloyd-if-4-3b-generator-of-pictures)

v1.0.0
------

- initial version


deepfloyd_if/modules/base.py
# -*- coding: utf-8 -*-
import os
import random
import platform
from datetime import datetime

import torch
import torchvision
import numpy as np
import matplotlib.pyplot as plt
import torchvision.transforms as T
from PIL import Image
from omegaconf import OmegaConf
from huggingface_hub import hf_hub_download
from accelerate.utils import set_module_tensor_to_device


from .. import utils
from ..model.respace import create_gaussian_diffusion
from .utils import load_model_weights, predict_proba, clip_process_generations


class IFBaseModule:

    stage = '-'

    available_models = []
    cpu_zero_emb = np.load(os.path.join(utils.RESOURCES_ROOT, 'zero_t5-v1_1-xxl_vector.npy'))
    cpu_zero_emb = torch.from_numpy(cpu_zero_emb)

    respacing_modes = {
        'fast27': '10,10,3,2,2',
        'smart27': '7,4,2,1,2,4,7',
        'smart50': '10,6,4,3,2,2,3,4,6,10',
        'smart100': '1,1,1,1,2,2,2,2,2,2,3,3,4,4,5,5,6,7,7,8,9,10,13',
        'smart185': '1,1,2,2,2,3,3,3,4,5,6,7,8,9,10,11,12,13,14,15,16,18,20',
        'super27': '1,1,1,1,1,1,1,2,5,13',  # for III super-res
        'super40': '2,2,2,2,2,2,3,4,6,15',  # for III super-res
        'super100': '4,4,6,6,8,8,10,10,14,30',  # for III super-res
    }

    wm_pil_img = Image.open(os.path.join(utils.RESOURCES_ROOT, 'wm.png'))

    try:
        import clip  # noqa
    except ModuleNotFoundError:
        print('Warning! You should install CLIP: "pip install git+https://github.com/openai/CLIP.git --no-deps"')
        raise

    clip_model, clip_preprocess = clip.load('ViT-L/14', device='cpu')
    clip_model.eval()

    cpu_w_weights, cpu_w_biases = load_model_weights(os.path.join(utils.RESOURCES_ROOT, 'w_head_v1.npz'))
    cpu_p_weights, cpu_p_biases = load_model_weights(os.path.join(utils.RESOURCES_ROOT, 'p_head_v1.npz'))
    w_threshold, p_threshold = 0.5, 0.5

    def __init__(self, dir_or_name, device, pil_img_size=256, cache_dir=None, hf_token=None):
        self.hf_token = hf_token
        self.cache_dir = cache_dir or os.path.expanduser('~/.cache/IF_')
        self.dir_or_name = dir_or_name
        self.conf = self.load_conf(dir_or_name) if not self.use_diffusers else None
        self.device = torch.device(device)
        self.zero_emb = self.cpu_zero_emb.clone().to(self.device)
        self.pil_img_size = pil_img_size

    @property
    def use_diffusers(self):
        return False

    def embeddings_to_image(
        self, t5_embs, low_res=None, *,
        style_t5_embs=None,
        positive_t5_embs=None,
        negative_t5_embs=None,
        batch_repeat=1,
        dynamic_thresholding_p=0.95,
        sample_loop='ddpm',
        sample_timestep_respacing='smart185',
        dynamic_thresholding_c=1.5,
        guidance_scale=7.0,
        aug_level=0.25,
        positive_mixer=0.15,
        free_us=None,
        free_ub=None,
        free_ur=0.2,
        blur_sigma=None,
        img_size=None,
        img_scale=4.0,
        aspect_ratio='1:1',
        progress=True,
        seed=None,
        sample_fn=None,
        support_noise=None,
        support_noise_less_qsample_steps=0,
        inpainting_mask=None,
        **kwargs,
    ):
        self._clear_cache()
        image_w, image_h = self._get_image_sizes(low_res, img_size, aspect_ratio, img_scale)
        diffusion = self.get_diffusion(sample_timestep_respacing)

        bs_scale = 2 if positive_t5_embs is None else 3

        def model_fn(x_t, ts, **kwargs):
            half = x_t[: len(x_t) // bs_scale]
            combined = torch.cat([half]*bs_scale, dim=0)
            model_out = self.model(combined, ts, free_us=free_us, free_ub=free_ub, free_ur=free_ur, **kwargs)
            eps, rest = model_out[:, :3], model_out[:, 3:]
            if bs_scale == 3:
                cond_eps, pos_cond_eps, uncond_eps = torch.split(eps, len(eps) // bs_scale, dim=0)
                half_eps = uncond_eps + guidance_scale * (
                    cond_eps * (1 - positive_mixer) + pos_cond_eps * positive_mixer - uncond_eps)
                pos_half_eps = uncond_eps + guidance_scale * (pos_cond_eps - uncond_eps)
                eps = torch.cat([half_eps, pos_half_eps, half_eps], dim=0)
            else:
                cond_eps, uncond_eps = torch.split(eps, len(eps) // bs_scale, dim=0)
                half_eps = uncond_eps + guidance_scale * (cond_eps - uncond_eps)
                eps = torch.cat([half_eps, half_eps], dim=0)
            return torch.cat([eps, rest], dim=1)

        seed = self.seed_everything(seed)

        text_emb = t5_embs.to(self.device, dtype=self.model.dtype).repeat(batch_repeat, 1, 1)
        batch_size = text_emb.shape[0] * batch_repeat

        if positive_t5_embs is not None:
            positive_t5_embs = positive_t5_embs.to(self.device, dtype=self.model.dtype).repeat(batch_repeat, 1, 1)

        if negative_t5_embs is not None:
            negative_t5_embs = negative_t5_embs.to(self.device, dtype=self.model.dtype).repeat(batch_repeat, 1, 1)

        timestep_text_emb = None
        if style_t5_embs is not None:
            list_timestep_text_emb = [
                style_t5_embs.to(self.device, dtype=self.model.dtype).repeat(batch_repeat, 1, 1),
            ]
            if positive_t5_embs is not None:
                list_timestep_text_emb.append(positive_t5_embs)
            if negative_t5_embs is not None:
                list_timestep_text_emb.append(negative_t5_embs)
            else:
                list_timestep_text_emb.append(
                    self.zero_emb.unsqueeze(0).repeat(batch_size, 1, 1).to(self.device, dtype=self.model.dtype))
            timestep_text_emb = torch.cat(list_timestep_text_emb, dim=0).to(self.device, dtype=self.model.dtype)

        metadata = {
            'seed': seed,
            'guidance_scale': guidance_scale,
            'dynamic_thresholding_p': dynamic_thresholding_p,
            'dynamic_thresholding_c': dynamic_thresholding_c,
            'batch_size': batch_size,
            'device_name': self.device_name,
            'img_size': [image_w, image_h],
            'sample_loop': sample_loop,
            'sample_timestep_respacing': sample_timestep_respacing,
            'stage': self.stage,
        }

        list_text_emb = [t5_embs.to(self.device)]
        if positive_t5_embs is not None:
            list_text_emb.append(positive_t5_embs.to(self.device))
        if negative_t5_embs is not None:
            list_text_emb.append(negative_t5_embs.to(self.device))
        else:
            list_text_emb.append(
                self.zero_emb.unsqueeze(0).repeat(batch_size, 1, 1).to(self.device, dtype=self.model.dtype))

        model_kwargs = dict(
            text_emb=torch.cat(list_text_emb, dim=0).to(self.device, dtype=self.model.dtype),
            timestep_text_emb=timestep_text_emb,
            use_cache=True,
        )
        if low_res is not None:
            if blur_sigma is not None:
                low_res = T.GaussianBlur(3, sigma=(blur_sigma, blur_sigma))(low_res)
            model_kwargs['low_res'] = torch.cat([low_res]*bs_scale, dim=0).to(self.device)
            model_kwargs['aug_level'] = aug_level

        if support_noise is None:
            noise = torch.randn(
                (batch_size * bs_scale, 3, image_h, image_w), device=self.device, dtype=self.model.dtype)
        else:
            assert support_noise_less_qsample_steps < len(diffusion.timestep_map) - 1
            assert support_noise.shape == (1, 3, image_h, image_w)
            q_sample_steps = torch.tensor([int(len(diffusion.timestep_map) - 1 - support_noise_less_qsample_steps)])
            support_noise = support_noise.cpu()
            noise = support_noise.clone()
            noise[inpainting_mask.cpu().bool() if inpainting_mask is not None else ...] = diffusion.q_sample(
                support_noise[inpainting_mask.cpu().bool() if inpainting_mask is not None else ...],
                q_sample_steps,
            )
            noise = noise.repeat(batch_size*bs_scale, 1, 1, 1).to(device=self.device, dtype=self.model.dtype)

        if inpainting_mask is not None:
            inpainting_mask = inpainting_mask.to(device=self.device, dtype=torch.long)

        if sample_loop == 'ddpm':
            with torch.no_grad():
                sample = diffusion.p_sample_loop(
                    model_fn,
                    (batch_size * bs_scale, 3, image_h, image_w),
                    noise=noise,
                    clip_denoised=True,
                    model_kwargs=model_kwargs,
                    dynamic_thresholding_p=dynamic_thresholding_p,
                    dynamic_thresholding_c=dynamic_thresholding_c,
                    inpainting_mask=inpainting_mask,
                    device=self.device,
                    progress=progress,
                    sample_fn=sample_fn,
                )[:batch_size]
        elif sample_loop == 'ddim':
            with torch.no_grad():
                sample = diffusion.ddim_sample_loop(
                    model_fn,
                    (batch_size * bs_scale, 3, image_h, image_w),
                    noise=noise,
                    clip_denoised=True,
                    model_kwargs=model_kwargs,
                    dynamic_thresholding_p=dynamic_thresholding_p,
                    dynamic_thresholding_c=dynamic_thresholding_c,
                    device=self.device,
                    progress=progress,
                    sample_fn=sample_fn,
                )[:batch_size]
        else:
            raise ValueError(f'Sample loop "{sample_loop}" doesnt support')

        sample = self.__validate_generations(sample)
        self._clear_cache()

        return sample, metadata

    def load_conf(self, dir_or_name, filename='config.yml'):
        path = self._get_path_or_download_file_from_hf(dir_or_name, filename)
        conf = OmegaConf.load(path)
        return conf

    def load_checkpoint(self, model, dir_or_name, filename='pytorch_model.bin'):
        path = self._get_path_or_download_file_from_hf(dir_or_name, filename)
        if os.path.exists(path):
            checkpoint = torch.load(path, map_location='cpu')
            param_device = 'cpu'
            for param_name, param in checkpoint.items():
                set_module_tensor_to_device(model, param_name, param_device, value=param)
        else:
            print(f'Warning! In directory "{dir_or_name}" filename "pytorch_model.bin" is not found.')
        return model

    def _get_path_or_download_file_from_hf(self, dir_or_name, filename):
        if dir_or_name in self.available_models:
            cache_dir = os.path.join(self.cache_dir, dir_or_name)
            hf_hub_download(repo_id=f'DeepFloyd/{dir_or_name}', filename=filename, cache_dir=cache_dir,
                            force_filename=filename, token=self.hf_token)
            return os.path.join(cache_dir, filename)
        else:
            return os.path.join(dir_or_name, filename)

    def get_diffusion(self, timestep_respacing):
        timestep_respacing = self.respacing_modes.get(timestep_respacing, timestep_respacing)
        diffusion = create_gaussian_diffusion(
            steps=1000,
            learn_sigma=True,
            sigma_small=False,
            noise_schedule='cosine',
            use_kl=False,
            predict_xstart=False,
            rescale_timesteps=True,
            rescale_learned_sigmas=True,
            timestep_respacing=timestep_respacing,
        )
        return diffusion

    @staticmethod
    def seed_everything(seed=None):
        if seed is None:
            seed = int((datetime.utcnow().timestamp() * 10 ** 6) % (2 ** 32 - 1))
        random.seed(seed)
        os.environ['PYTHONHASHSEED'] = str(seed)
        np.random.seed(seed)
        torch.manual_seed(seed)
        torch.cuda.manual_seed(seed)
        torch.backends.cudnn.deterministic = True
        torch.backends.cudnn.benchmark = True
        return seed

    def device_name(self):
        if self.device.type == 'cpu':
            return 'cpu_' + str(platform.processor())
        if self.device.type == 'cuda':
            return torch.cuda.get_device_name(self.device)
        return '-'

    def to_images(self, generations, disable_watermark=False):
        bs, c, h, w = generations.shape
        coef = min(h / self.pil_img_size, w / self.pil_img_size)
        img_h, img_w = (int(h / coef), int(w / coef)) if coef < 1 else (h, w)

        S1, S2 = 1024 ** 2, img_w * img_h
        K = (S2 / S1) ** 0.5
        wm_size, wm_x, wm_y = int(K * 62), img_w - int(14 * K), img_h - int(14 * K)

        wm_img = self.wm_pil_img.resize(
            (wm_size, wm_size), getattr(Image, 'Resampling', Image).BICUBIC, reducing_gap=None)

        pil_images = []
        for image in ((generations + 1) * 127.5).round().clamp(0, 255).to(torch.uint8).cpu():
            pil_img = torchvision.transforms.functional.to_pil_image(image).convert('RGB')
            pil_img = pil_img.resize((img_w, img_h), getattr(Image, 'Resampling', Image).NEAREST)
            if not disable_watermark:
                pil_img.paste(wm_img, box=(wm_x - wm_size, wm_y - wm_size, wm_x, wm_y), mask=wm_img.split()[-1])
            pil_images.append(pil_img)
        return pil_images

    def show(self, pil_images, nrow=None, size=10):
        if nrow is None:
            nrow = round(len(pil_images)**0.5)

        imgs = torchvision.utils.make_grid(utils.pil_list_to_torch_tensors(pil_images), nrow=nrow)
        if not isinstance(imgs, list):
            imgs = [imgs.cpu()]

        fix, axs = plt.subplots(ncols=len(imgs), squeeze=False, figsize=(size, size))
        for i, img in enumerate(imgs):
            img = img.detach()
            img = torchvision.transforms.functional.to_pil_image(img)
            axs[0, i].imshow(np.asarray(img))
            axs[0, i].set(xticklabels=[], yticklabels=[], xticks=[], yticks=[])

        fix.show()
        plt.show()

    def _clear_cache(self):
        self.model.cache = None

    def _get_image_sizes(self, low_res, img_size, aspect_ratio, img_scale):
        if low_res is not None:
            bs, c, h, w = low_res.shape
            image_h, image_w = int((h*img_scale)//32)*32, int((w*img_scale//32))*32
        else:
            scale_w, scale_h = aspect_ratio.split(':')
            scale_w, scale_h = int(scale_w), int(scale_h)
            coef = scale_w / scale_h
            image_h, image_w = img_size, img_size
            if coef >= 1:
                image_w = int(round(img_size/8 * coef) * 8)
            else:
                image_h = int(round(img_size/8 / coef) * 8)

        assert image_h % 8 == 0
        assert image_w % 8 == 0

        return image_w, image_h

    def __validate_generations(self, generations):
        with torch.no_grad():
            imgs = clip_process_generations(generations)
            image_features = self.clip_model.encode_image(imgs.to('cpu'))
            image_features = image_features.detach().cpu().numpy().astype(np.float16)
            p_pred = predict_proba(image_features, self.cpu_p_weights, self.cpu_p_biases)
            w_pred = predict_proba(image_features, self.cpu_w_weights, self.cpu_w_biases)
            query = p_pred > self.p_threshold
            if query.sum() > 0:
                generations[query] = T.GaussianBlur(99, sigma=(100.0, 100.0))(generations[query])
            query = w_pred > self.w_threshold
            if query.sum() > 0:
                generations[query] = T.GaussianBlur(99, sigma=(100.0, 100.0))(generations[query])
        return generations
